src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,411 @@
+                        _stateProvider.DeleteAccount(toBeDestroyed);
+                    }
+
+                    statusCode = StatusCode.Success;
+                }
+
+                spentGas = Refund(gasLimit, unspentGas, substate, sender, gasPrice, spec);
+            }
+            catch (Exception ex) when (ex is EvmException || ex is OverflowException) // TODO: OverflowException? still needed? hope not
+            {
+                if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose($"EVM EXCEPTION: {ex.GetType().Name}");
+                _stateProvider.Restore(stateSnapshot);
+                _storageProvider.Restore(storageSnapshot);
+            }
+
+            if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Gas spent: " + spentGas);
+
+            Address gasBeneficiary = stateUpdate.GasBeneficiary;
+            if (statusCode == StatusCode.Failure || !(substate?.DestroyList.Contains(gasBeneficiary) ?? false))
+            {
+                if (!_stateProvider.AccountExists(gasBeneficiary))
+                {
+                    _stateProvider.CreateAccount(gasBeneficiary, (ulong) spentGas * gasPrice);
+                }
+                else
+                {
+                    _stateProvider.AddToBalance(gasBeneficiary, (ulong) spentGas * gasPrice, spec);
+                }
+            }
+
+            if (!readOnly)
+            {
+                _storageProvider.Commit(txTracer.IsTracingState ? txTracer : null);
+                _stateProvider.Commit(spec, txTracer.IsTracingState ? txTracer : null);
+            }
+            else
+            {
+                _storageProvider.Reset();
+                _stateProvider.Reset();
+            }
+
+            if (!readOnly)
+            {
+                stateUpdate.GasUsed += spentGas;
+            }
+
+            if (txTracer.IsTracingReceipt)
+            {
+                if (statusCode == StatusCode.Failure)
+                {
+                    txTracer.MarkAsFailed(recipient, spentGas, (substate?.ShouldRevert ?? false) ? substate.Output : Bytes.Empty, substate?.Error);
+                }
+                else
+                {
+                    txTracer.MarkAsSuccess(recipient, spentGas, substate.Output, substate.Logs.Any() ? substate.Logs.ToArray() : LogEntry.EmptyLogs);
+                }
+            }
+        }
+
+        private long Refund(long gasLimit, long unspentGas, TransactionSubstate substate, Address sender, UInt256 gasPrice, IReleaseSpec spec)
+        {
+            long spentGas = gasLimit;
+            if (!substate.IsError)
+            {
+                spentGas -= unspentGas;
+                long refund = substate.ShouldRevert ? 0 : Math.Min(spentGas / 2L, substate.Refund + substate.DestroyList.Count * RefundOf.Destroy);
+
+                if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Refunding unused gas of " + unspentGas + " and refund of " + refund);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)